# FE520C Final Exam Coding Part

### Question 1 (25pt)

Design a Portfolio Management System using Python classes that can manage multiple financial assets and calculate portfolio performance. Implement the following classes:

1. `Asset`: Represents a financial asset in the portfolio. Each asset should have the following attributes: `ticker`, `quantity`, `purchase_price`, and `current_price`.

2. `Portfolio`: Represents a collection of financial assets. It should have the following functionality:

   - `add_asset(self, asset: Asset)`: Adds an asset to the portfolio.
   - `update_price(self, ticker: str, new_price: float)`: Updates the current price of an asset in the portfolio.
   - `total_investment(self)`: Returns the total investment made in the portfolio. (quantity * purchase_price)
   - `total_current_value(self)`: Returns the total current value of the portfolio. (quantity * current_price)
   - `asset_return(self, ticker: str)`: Returns the individual return of an asset in the portfolio.
   - `total_return(self)`: Returns the total return of the portfolio.
   - `weighted_return(self)`: Returns the weighted return of the portfolio.

Assume that the individual return of an asset is calculated as `(current_price - purchase_price) / purchase_price` and the weighted return of the portfolio is the average of individual asset returns, weighted by their current value in the portfolio.

In [3]:
class Asset:
    def __init__(self, ticker: str, quantity: int, purchase_price: float, current_price: float):
        self.ticker = ticker
        self.quantity = quantity
        self.purchase_price = purchase_price
        self.current_price = current_price

    def individual_return(self) -> float:
        return (self.current_price - self.purchase_price) / self.purchase_price


class Portfolio:
    def __init__(self):
        self.assets = {}

    def add_asset(self, asset: Asset):
        self.assets[asset.ticker] = asset

    def update_price(self, ticker: str, new_price: float):
        self.assets[ticker].current_price = new_price

    def total_investment(self) -> float:
        return sum(asset.quantity * asset.purchase_price for asset in self.assets.values())

    def total_current_value(self) -> float:
        return sum(asset.quantity * asset.current_price for asset in self.assets.values())

    def asset_return(self, ticker: str) -> float:
        return self.assets[ticker].individual_return()

    def total_return(self) -> float:
        return (self.total_current_value() - self.total_investment()) / self.total_investment()

    def weighted_return(self) -> float:
        total_value = self.total_current_value()
        weighted_returns = [asset.individual_return() * asset.quantity * asset.current_price / total_value
                            for asset in self.assets.values()]
        return sum(weighted_returns)

In [2]:
# Example:

asset1 = Asset("AAPL", 10, 150, 170)
asset2 = Asset("GOOG", 5, 1000, 1100)
asset3 = Asset("TSLA", 8, 300, 250)

portfolio = Portfolio()
portfolio.add_asset(asset1)
portfolio.add_asset(asset2)
portfolio.add_asset(asset3)

print("Total Investment:", portfolio.total_investment())
print("Total Current Value:", portfolio.total_current_value())
print("AAPL Return:", portfolio.asset_return("AAPL"))
print("Total Portfolio Return:", portfolio.total_return())
print("Weighted Portfolio Return:", portfolio.weighted_return())

portfolio.update_price("AAPL", 180)
print("Updated AAPL Return:", portfolio.asset_return("AAPL"))
print("Updated Total Portfolio Return:", portfolio.total_return())
print("Updated Weighted Portfolio Return:", portfolio.weighted_return())

Total Investment: 8900
Total Current Value: 9200
AAPL Return: 0.13333333333333333
Total Portfolio Return: 0.033707865168539325
Weighted Portfolio Return: 0.04818840579710145
Updated AAPL Return: 0.2
Updated Total Portfolio Return: 0.0449438202247191
Updated Weighted Portfolio Return: 0.06200716845878137


Expected solution:

Total Investment: 8900

Total Current Value: 9200

AAPL Return: 0.13333333333333333

Total Portfolio Return: 0.033707865168539325

Weighted Portfolio Return: 0.04818840579710145

Updated AAPL Return: 0.2

Updated Total Portfolio Return: 0.0449438202247191

Updated Weighted Portfolio Return: 0.06200716845878137

### Question 2 (25pt)

Problem:

You are given a list of integers, and you need to calculate the sum of squares of the even numbers and the sum of cubes of the odd numbers.

Write a Python function called `calculate_sums(lst: List[int]) -> Tuple[int, int]` that takes a list of integers as input and returns a tuple with two integers. The first integer in the tuple is the sum of squares of the even numbers, and the second integer is the sum of cubes of the odd numbers. (12pt)

In [4]:
from typing import List, Tuple

def calculate_sums(lst: List[int]) -> Tuple[int, int]:
    sum_of_squares = 0
    sum_of_cubes = 0
    
    for num in lst:
        if num % 2 == 0:
            sum_of_squares += num ** 2
        else:
            sum_of_cubes += num ** 3
    
    return (sum_of_squares, sum_of_cubes)    

In [5]:
# Example usage:
lst = [1, 2, 3, 4, 5, 6]
result = calculate_sums(lst)
print(result)

(56, 153)


(2) Create a class `Circle` with the following requirements:

1. It should have an instance variable `radius` with a default value of 1.
2. The class should have a method `area()` that returns the area of the circle.
3. The class should have a method `circumference()` that returns the circumference of the circle.

Note: You can use the value of pi as 3.14159 or import it from the `math` module. (12pt)

In [6]:
import math

class Circle:
    def __init__(self, radius=1):
        self.radius = radius
        
    def area(self):
        return math.pi * (self.radius ** 2)
    
    def circumference(self):
        return 2 * math.pi * self.radius

In [7]:
# Example usage:
circle = Circle(5)
print("Area:", circle.area())
print("Circumference:", circle.circumference())

Area: 78.53981633974483
Circumference: 31.41592653589793
